building automaticaly generated code with -Wconversion -Wpedantic

When I am building the code generated with cddl-gen with the flags -Wconversion -Wpedantic I am getting many warnings, see below. Is it possible to update the code so that it can complete with -Wconversion -Wpedantic without warnings?

You can reproduce the warnings by:

```
git clone --recurse-submodules -j8 git@github.com:Fraunhofer-AISEC/uoscore-uedhoc.git
cd uoscore-uedhoc
git checkout adb019f474142a267d22d6e0b901947e84cedbce
make
```




```
In file included from modules/edhoc/src/../edhoc.h:17,
                 from modules/edhoc/src/initiator.c:12:
modules/edhoc/src/initiator.c: In function ‘edhoc_initiator_run’:
modules/edhoc/src/initiator.c:305:35: warning: conversion from ‘uint32_t’ {aka ‘unsigned int’} to ‘uint8_t’ {aka ‘unsigned char’} may change value [-Wconversion]
  305 |    CIPHERTEXT4, &suite, prk_4x3m, prk_4x3m_len, th4,
      |                                   ^~~~~~~~~~~~
modules/edhoc/src/../../common/inc/oscore_edhoc_error.h:86:17: note: in definition of macro ‘TRY’
   86 |   int retval = (x);                                                  \
      |                 ^
modules/edhoc/src/initiator.c:306:4: warning: conversion from ‘uint32_t’ {aka ‘unsigned int’} to ‘uint8_t’ {aka ‘unsigned char’} may change value [-Wconversion]
  306 |    th4_len, ciphertext_4, ciphertext_4_len, NULL, 0, NULL,
      |    ^~~~~~~
modules/edhoc/src/../../common/inc/oscore_edhoc_error.h:86:17: note: in definition of macro ‘TRY’
   86 |   int retval = (x);                                                  \
      |                 ^
[Compile] modules/edhoc/cbor/encode_int_type.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_int_type.c:11:
modules/edhoc/cbor/encode_int_type.c: In function ‘cbor_encode_int_type_i’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_int_type.c:46:22: note: in expansion of macro ‘MIN’
   46 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_th4.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_th4.c:11:
modules/edhoc/cbor/encode_th4.c: In function ‘cbor_encode_th4’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_th4.c:47:22: note: in expansion of macro ‘MIN’
   47 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_cert.c
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_cert.c:11:
modules/edhoc/cbor/decode_cert.c: In function ‘cbor_decode_cert’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_cert.c:54:22: note: in expansion of macro ‘MIN’
   54 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_bstr_type.c
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_bstr_type.c:11:
modules/edhoc/cbor/decode_bstr_type.c: In function ‘cbor_decode_bstr_type_b_str’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_bstr_type.c:46:22: note: in expansion of macro ‘MIN’
   46 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_th3.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_th3.c:11:
modules/edhoc/cbor/encode_th3.c: In function ‘cbor_encode_th3’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_th3.c:47:22: note: in expansion of macro ‘MIN’
   47 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_data_2.c
modules/edhoc/cbor/encode_data_2.c: In function ‘encode_data_2’:
modules/edhoc/cbor/encode_data_2.c:39:71: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   39 |  bool tmp_result = (((present_encode(&((*input)._data_2_C_I_present), (void *)encode_repeated_data_2_C_I, state, (&(*input)._data_2_C_I))
      |                                                                       ^
modules/edhoc/cbor/encode_data_2.c:39:71: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   39 |  bool tmp_result = (((present_encode(&((*input)._data_2_C_I_present), (void *)encode_repeated_data_2_C_I, state, (&(*input)._data_2_C_I))
      |                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_data_2.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_data_2.c:11:
modules/edhoc/cbor/encode_data_2.c: In function ‘cbor_encode_data_2’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_data_2.c:65:22: note: in expansion of macro ‘MIN’
   65 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_id_cred_x.c
modules/edhoc/cbor/encode_id_cred_x.c: In function ‘encode_id_cred_x_map’:
modules/edhoc/cbor/encode_id_cred_x.c:83:120: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   83 |  bool tmp_result = (((map_start_encode(state, 4) && (int_res = (present_encode(&((*input)._id_cred_x_map_kid_present), (void *)encode_repeated_id_cred_x_map_kid, state, (&(*input)._id_cred_x_map_kid))
      |                                                                                                                        ^
modules/edhoc/cbor/encode_id_cred_x.c:83:120: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   83 |  bool tmp_result = (((map_start_encode(state, 4) && (int_res = (present_encode(&((*input)._id_cred_x_map_kid_present), (void *)encode_repeated_id_cred_x_map_kid, state, (&(*input)._id_cred_x_map_kid))
      |                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/encode_id_cred_x.c:84:64: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   84 |  && present_encode(&((*input)._id_cred_x_map_x5chain_present), (void *)encode_repeated_id_cred_x_map_x5chain, state, (&(*input)._id_cred_x_map_x5chain))
      |                                                                ^
modules/edhoc/cbor/encode_id_cred_x.c:84:64: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   84 |  && present_encode(&((*input)._id_cred_x_map_x5chain_present), (void *)encode_repeated_id_cred_x_map_x5chain, state, (&(*input)._id_cred_x_map_x5chain))
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/encode_id_cred_x.c:85:60: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   85 |  && present_encode(&((*input)._id_cred_x_map_x5t_present), (void *)encode_repeated_id_cred_x_map_x5t, state, (&(*input)._id_cred_x_map_x5t))
      |                                                            ^
modules/edhoc/cbor/encode_id_cred_x.c:85:60: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   85 |  && present_encode(&((*input)._id_cred_x_map_x5t_present), (void *)encode_repeated_id_cred_x_map_x5t, state, (&(*input)._id_cred_x_map_x5t))
      |                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/encode_id_cred_x.c:86:60: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   86 |  && present_encode(&((*input)._id_cred_x_map_x5u_present), (void *)encode_repeated_id_cred_x_map_x5u, state, (&(*input)._id_cred_x_map_x5u))), ((map_end_encode(state, 4)) && int_res)))));
      |                                                            ^
modules/edhoc/cbor/encode_id_cred_x.c:86:60: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   86 |  && present_encode(&((*input)._id_cred_x_map_x5u_present), (void *)encode_repeated_id_cred_x_map_x5u, state, (&(*input)._id_cred_x_map_x5u))), ((map_end_encode(state, 4)) && int_res)))));
      |                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_id_cred_x.c:11:
modules/edhoc/cbor/encode_id_cred_x.c: In function ‘cbor_encode_id_cred_x_map’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_id_cred_x.c:108:22: note: in expansion of macro ‘MIN’
  108 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_message_error.c
modules/edhoc/cbor/encode_message_error.c: In function ‘encode_repeated_message_error_SUITES_R’:
modules/edhoc/cbor/encode_message_error.c:40:205: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   40 | age_error_SUITES_R_choice == _SUITES_R__supported) ? ((list_start_encode(state, 10) && (int_res = (multi_encode(2, 10, &(*input)._SUITES_R__supported_supported_count, (void *)intx32_encode, state, (&(*input)._SUITES_R__supported_supported), sizeof(int32_t))), ((list_end_encode(state, 10)) && int_res))))
      |                                                                                                                                                                        ^

modules/edhoc/cbor/encode_message_error.c:40:205: warning: ISO C forbids passing argument 4 of ‘multi_encode’ between function pointer and ‘void *’ [-Wpedantic]
   40 | age_error_SUITES_R_choice == _SUITES_R__supported) ? ((list_start_encode(state, 10) && (int_res = (multi_encode(2, 10, &(*input)._SUITES_R__supported_supported_count, (void *)intx32_encode, state, (&(*input)._SUITES_R__supported_supported), sizeof(int32_t))), ((list_end_encode(state, 10)) && int_res))))
      |                                                                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~

In file included from modules/edhoc/cbor/encode_message_error.c:11:
externals/cddl-gen/include/cbor_encode.h:137:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  137 |   cbor_encoder_t encoder, cbor_state_t *state, const void *input,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/encode_message_error.c: In function ‘encode_message_error’:
modules/edhoc/cbor/encode_message_error.c:55:78: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   55 |  bool tmp_result = (((present_encode(&((*input)._message_error_C_x_present), (void *)encode_repeated_message_error_C_x, state, (&(*input)._message_error_C_x))
      |                                                                              ^
modules/edhoc/cbor/encode_message_error.c:55:78: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   55 |  bool tmp_result = (((present_encode(&((*input)._message_error_C_x_present), (void *)encode_repeated_message_error_C_x, state, (&(*input)._message_error_C_x))
      |                                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_message_error.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/encode_message_error.c:57:65: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   57 |  && present_encode(&((*input)._message_error_SUITES_R_present), (void *)encode_repeated_message_error_SUITES_R, state, (&(*input)._message_error_SUITES_R)))));
      |                                                                 ^
modules/edhoc/cbor/encode_message_error.c:57:65: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   57 |  && present_encode(&((*input)._message_error_SUITES_R_present), (void *)encode_repeated_message_error_SUITES_R, state, (&(*input)._message_error_SUITES_R)))));
      |                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_message_error.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_message_error.c:11:
modules/edhoc/cbor/encode_message_error.c: In function ‘cbor_encode_message_error’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_message_error.c:79:22: note: in expansion of macro ‘MIN’
   79 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_message_1.c
modules/edhoc/cbor/decode_message_1.c: In function ‘decode_message_1’:
modules/edhoc/cbor/decode_message_1.c:26:150: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   26 |  && ((union_start_code(state) && (int_res = ((((list_start_decode(state) && (int_res = (multi_decode(2, 10, &(*result)._SUITES_I__suite_suite_count, (void *)intx32_decode, state, (&(*result)._SUITES_I__suite_suite), sizeof(int32_t))), ((list_end_decode(state)) && int_res)))) && (((*result)._message_1_SUITES_I_choice = _SUITES_I__suite) || 1))
      |                                                                                                                                                      ^
modules/edhoc/cbor/decode_message_1.c:26:150: warning: ISO C forbids passing argument 4 of ‘multi_decode’ between function pointer and ‘void *’ [-Wpedantic]
   26 |  && ((union_start_code(state) && (int_res = ((((list_start_decode(state) && (int_res = (multi_decode(2, 10, &(*result)._SUITES_I__suite_suite_count, (void *)intx32_decode, state, (&(*result)._SUITES_I__suite_suite), sizeof(int32_t))), ((list_end_decode(state)) && int_res)))) && (((*result)._message_1_SUITES_I_choice = _SUITES_I__suite) || 1))
      |                                                                                                                                                      ^~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_message_1.c:11:
externals/cddl-gen/include/cbor_decode.h:221:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  221 |   cbor_decoder_t decoder, cbor_state_t *state, void *result,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_message_1.c:31:59: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   31 |  && present_decode(&((*result)._message_1_ead_1_present), (void *)bstrx_decode, state, (&(*result)._message_1_ead_1)))));
      |                                                           ^
modules/edhoc/cbor/decode_message_1.c:31:59: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   31 |  && present_decode(&((*result)._message_1_ead_1_present), (void *)bstrx_decode, state, (&(*result)._message_1_ead_1)))));
      |                                                           ^~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_message_1.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_message_1.c:11:
modules/edhoc/cbor/decode_message_1.c: In function ‘cbor_decode_message_1’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_message_1.c:53:22: note: in expansion of macro ‘MIN’
   53 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_message_2.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_message_2.c:11:
modules/edhoc/cbor/encode_message_2.c: In function ‘cbor_encode_m2’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_message_2.c:49:22: note: in expansion of macro ‘MIN’
   49 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_th2.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_th2.c:11:
modules/edhoc/cbor/encode_th2.c: In function ‘cbor_encode_th2’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_th2.c:50:22: note: in expansion of macro ‘MIN’
   50 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_sig_structure.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_sig_structure.c:11:
modules/edhoc/cbor/encode_sig_structure.c: In function ‘cbor_encode_sig_structure’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_sig_structure.c:50:22: note: in expansion of macro ‘MIN’
   50 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_info.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_info.c:11:
modules/edhoc/cbor/encode_info.c: In function ‘cbor_encode_info’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_info.c:49:22: note: in expansion of macro ‘MIN’
   49 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_bstr_type.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_bstr_type.c:11:
modules/edhoc/cbor/encode_bstr_type.c: In function ‘cbor_encode_bstr_type_b_str’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_bstr_type.c:46:22: note: in expansion of macro ‘MIN’
   46 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_enc_structure.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_enc_structure.c:11:
modules/edhoc/cbor/encode_enc_structure.c: In function ‘cbor_encode_enc_structure’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_enc_structure.c:49:22: note: in expansion of macro ‘MIN’
   49 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_message_1.c
modules/edhoc/cbor/encode_message_1.c: In function ‘encode_message_1’:
modules/edhoc/cbor/encode_message_1.c:26:173: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   26 | (((*input)._message_1_SUITES_I_choice == _SUITES_I__suite) ? ((list_start_encode(state, 10) && (int_res = (multi_encode(2, 10, &(*input)._SUITES_I__suite_suite_count, (void *)intx32_encode, state, (&(*input)._SUITES_I__suite_suite), sizeof(int32_t))), ((list_end_encode(state, 10)) && int_res))))
      |                                                                                                                                                                        ^

modules/edhoc/cbor/encode_message_1.c:26:173: warning: ISO C forbids passing argument 4 of ‘multi_encode’ between function pointer and ‘void *’ [-Wpedantic]
   26 | (((*input)._message_1_SUITES_I_choice == _SUITES_I__suite) ? ((list_start_encode(state, 10) && (int_res = (multi_encode(2, 10, &(*input)._SUITES_I__suite_suite_count, (void *)intx32_encode, state, (&(*input)._SUITES_I__suite_suite), sizeof(int32_t))), ((list_end_encode(state, 10)) && int_res))))
      |                                                                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~

In file included from modules/edhoc/cbor/encode_message_1.c:11:
externals/cddl-gen/include/cbor_encode.h:137:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  137 |   cbor_encoder_t encoder, cbor_state_t *state, const void *input,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/encode_message_1.c:33:58: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   33 |  && present_encode(&((*input)._message_1_ead_1_present), (void *)bstrx_encode, state, (&(*input)._message_1_ead_1)))));
      |                                                          ^
modules/edhoc/cbor/encode_message_1.c:33:58: warning: ISO C forbids passing argument 2 of ‘present_encode’ between function pointer and ‘void *’ [-Wpedantic]
   33 |  && present_encode(&((*input)._message_1_ead_1_present), (void *)bstrx_encode, state, (&(*input)._message_1_ead_1)))));
      |                                                          ^~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/encode_message_1.c:11:
externals/cddl-gen/include/cbor_encode.h:141:18: note: expected ‘_Bool (*)(cbor_state_t *, const void *)’ {aka ‘_Bool (*)(struct <anonymous> *, const void *)’} but argument is of type ‘void *’
  141 |   cbor_encoder_t encoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_message_1.c:11:
modules/edhoc/cbor/encode_message_1.c: In function ‘cbor_encode_message_1’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_message_1.c:55:22: note: in expansion of macro ‘MIN’
   55 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/encode_message_3.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/edhoc/cbor/encode_message_3.c:11:
modules/edhoc/cbor/encode_message_3.c: In function ‘cbor_encode_m3_CIPHERTEXT_3’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/encode_message_3.c:46:22: note: in expansion of macro ‘MIN’
   46 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_id_cred_x.c
modules/edhoc/cbor/decode_id_cred_x.c: In function ‘decode_id_cred_x_map’:
modules/edhoc/cbor/decode_id_cred_x.c:83:118: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   83 |  bool tmp_result = (((map_start_decode(state) && (int_res = (present_decode(&((*result)._id_cred_x_map_kid_present), (void *)decode_repeated_id_cred_x_map_kid, state, (&(*result)._id_cred_x_map_kid))
      |                                                                                                                      ^
modules/edhoc/cbor/decode_id_cred_x.c:83:118: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   83 |  bool tmp_result = (((map_start_decode(state) && (int_res = (present_decode(&((*result)._id_cred_x_map_kid_present), (void *)decode_repeated_id_cred_x_map_kid, state, (&(*result)._id_cred_x_map_kid))
      |                                                                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_id_cred_x.c:84:65: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   84 |  && present_decode(&((*result)._id_cred_x_map_x5chain_present), (void *)decode_repeated_id_cred_x_map_x5chain, state, (&(*result)._id_cred_x_map_x5chain))
      |                                                                 ^
modules/edhoc/cbor/decode_id_cred_x.c:84:65: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   84 |  && present_decode(&((*result)._id_cred_x_map_x5chain_present), (void *)decode_repeated_id_cred_x_map_x5chain, state, (&(*result)._id_cred_x_map_x5chain))
      |                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_id_cred_x.c:85:61: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   85 |  && present_decode(&((*result)._id_cred_x_map_x5t_present), (void *)decode_repeated_id_cred_x_map_x5t, state, (&(*result)._id_cred_x_map_x5t))
      |                                                             ^
modules/edhoc/cbor/decode_id_cred_x.c:85:61: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   85 |  && present_decode(&((*result)._id_cred_x_map_x5t_present), (void *)decode_repeated_id_cred_x_map_x5t, state, (&(*result)._id_cred_x_map_x5t))
      |                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_id_cred_x.c:86:61: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   86 |  && present_decode(&((*result)._id_cred_x_map_x5u_present), (void *)decode_repeated_id_cred_x_map_x5u, state, (&(*result)._id_cred_x_map_x5u))), ((map_end_decode(state)) && int_res)))));
      |                                                             ^
modules/edhoc/cbor/decode_id_cred_x.c:86:61: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   86 |  && present_decode(&((*result)._id_cred_x_map_x5u_present), (void *)decode_repeated_id_cred_x_map_x5u, state, (&(*result)._id_cred_x_map_x5u))), ((map_end_decode(state)) && int_res)))));
      |                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_id_cred_x.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_id_cred_x.c:11:
modules/edhoc/cbor/decode_id_cred_x.c: In function ‘cbor_decode_id_cred_x_map’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_id_cred_x.c:108:22: note: in expansion of macro ‘MIN’
  108 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_message_2.c
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_message_2.c:11:
modules/edhoc/cbor/decode_message_2.c: In function ‘cbor_decode_m2’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_message_2.c:49:22: note: in expansion of macro ‘MIN’
   49 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_message_3.c
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_message_3.c:11:
modules/edhoc/cbor/decode_message_3.c: In function ‘cbor_decode_m3_CIPHERTEXT_3’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_message_3.c:46:22: note: in expansion of macro ‘MIN’
   46 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/edhoc/cbor/decode_plaintext.c
modules/edhoc/cbor/decode_plaintext.c: In function ‘decode_map’:
modules/edhoc/cbor/decode_plaintext.c:83:108: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   83 |  bool tmp_result = (((map_start_decode(state) && (int_res = (present_decode(&((*result)._map_kid_present), (void *)decode_repeated_map_kid, state, (&(*result)._map_kid))
      |                                                                                                            ^
modules/edhoc/cbor/decode_plaintext.c:83:108: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   83 |  bool tmp_result = (((map_start_decode(state) && (int_res = (present_decode(&((*result)._map_kid_present), (void *)decode_repeated_map_kid, state, (&(*result)._map_kid))
      |                                                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_plaintext.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_plaintext.c:84:55: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   84 |  && present_decode(&((*result)._map_x5chain_present), (void *)decode_repeated_map_x5chain, state, (&(*result)._map_x5chain))
      |                                                       ^
modules/edhoc/cbor/decode_plaintext.c:84:55: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   84 |  && present_decode(&((*result)._map_x5chain_present), (void *)decode_repeated_map_x5chain, state, (&(*result)._map_x5chain))
      |                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_plaintext.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_plaintext.c:85:51: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   85 |  && present_decode(&((*result)._map_x5t_present), (void *)decode_repeated_map_x5t, state, (&(*result)._map_x5t))
      |                                                   ^
modules/edhoc/cbor/decode_plaintext.c:85:51: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   85 |  && present_decode(&((*result)._map_x5t_present), (void *)decode_repeated_map_x5t, state, (&(*result)._map_x5t))
      |                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_plaintext.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_plaintext.c:86:51: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
   86 |  && present_decode(&((*result)._map_x5u_present), (void *)decode_repeated_map_x5u, state, (&(*result)._map_x5u))), ((map_end_decode(state)) && int_res)))));
      |                                                   ^
modules/edhoc/cbor/decode_plaintext.c:86:51: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
   86 |  && present_decode(&((*result)._map_x5u_present), (void *)decode_repeated_map_x5u, state, (&(*result)._map_x5u))), ((map_end_decode(state)) && int_res)))));
      |                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_plaintext.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
modules/edhoc/cbor/decode_plaintext.c: In function ‘decode_plaintext’:
modules/edhoc/cbor/decode_plaintext.c:104:58: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
  104 |  && present_decode(&((*result)._plaintext_AD_x_present), (void *)bstrx_decode, state, (&(*result)._plaintext_AD_x)))));
      |                                                          ^
modules/edhoc/cbor/decode_plaintext.c:104:58: warning: ISO C forbids passing argument 2 of ‘present_decode’ between function pointer and ‘void *’ [-Wpedantic]
  104 |  && present_decode(&((*result)._plaintext_AD_x_present), (void *)bstrx_decode, state, (&(*result)._plaintext_AD_x)))));
      |                                                          ^~~~~~~~~~~~~~~~~~~~
In file included from modules/edhoc/cbor/decode_plaintext.c:11:
externals/cddl-gen/include/cbor_decode.h:225:18: note: expected ‘_Bool (*)(cbor_state_t *, void *)’ {aka ‘_Bool (*)(struct <anonymous> *, void *)’} but argument is of type ‘void *’
  225 |   cbor_decoder_t decoder,
      |   ~~~~~~~~~~~~~~~^~~~~~~
In file included from externals/cddl-gen/include/cbor_decode.h:12,
                 from modules/edhoc/cbor/decode_plaintext.c:11:
modules/edhoc/cbor/decode_plaintext.c: In function ‘cbor_decode_plaintext’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/edhoc/cbor/decode_plaintext.c:126:22: note: in expansion of macro ‘MIN’
  126 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/oscore/src/security_context.c
[Compile] modules/oscore/src/coap2oscore.c
[Compile] modules/oscore/src/oscore2coap.c
[Compile] modules/oscore/src/osocre_hkdf_info.c
[Compile] modules/oscore/src/oscore_cose.c
[Compile] modules/oscore/src/option.c
[Compile] modules/oscore/src/aad.c
[Compile] modules/oscore/src/nonce.c
[Compile] modules/oscore/src/coap.c
[Compile] modules/oscore/cbor/enc_structure.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/oscore/cbor/enc_structure.c:11:
modules/oscore/cbor/enc_structure.c: In function ‘cbor_encode_enc_structure’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/oscore/cbor/enc_structure.c:49:22: note: in expansion of macro ‘MIN’
   49 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/oscore/cbor/info.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/oscore/cbor/info.c:11:
modules/oscore/cbor/info.c: In function ‘cbor_encode_oscore_info’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/oscore/cbor/info.c:55:22: note: in expansion of macro ‘MIN’
   55 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Compile] modules/oscore/cbor/aad_array.c
In file included from externals/cddl-gen/include/cbor_encode.h:12,
                 from modules/oscore/cbor/aad_array.c:11:
modules/oscore/cbor/aad_array.c: In function ‘cbor_encode_aad_array’:
externals/cddl-gen/include/cbor_common.h:46:19: warning: conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Wconversion]
   46 | #define MIN(a, b) (((a) < (b)) ? (a) : (b))
      |                   ^
modules/oscore/cbor/aad_array.c:53:22: note: in expansion of macro ‘MIN’
   53 |   *payload_len_out = MIN(payload_len,
      |                      ^~~
[Link (Static)]```

--- a/include/zcbor_common.h
+++ b/include/zcbor_common.h
@@ -120,7 +120,7 @@ do {\
 #define ZCBOR_VALUE_IS_8_BYTES 27 ///! The next 8 bytes contain the value.
 #define ZCBOR_VALUE_IS_INDEFINITE_LENGTH 31 ///! The list or map has indefinite length, and will instead be terminated by a 0xFF token.
 
-#define ZCBOR_BOOL_TO_PRIM 20 ///! In CBOR, false/true have the values 20/21
+#define ZCBOR_BOOL_TO_PRIM ((uint8_t)20) ///! In CBOR, false/true have the values 20/21
 
 #define ZCBOR_FLAG_RESTORE 1UL ///! Restore from the backup. Overwrite the current state with the state from the backup.
 #define ZCBOR_FLAG_CONSUME 2UL ///! Consume the backup. Remove the backup from the stack of backups.

--- a/src/zcbor_decode.c
+++ b/src/zcbor_decode.c
@@ -22,7 +22,7 @@ static uint_fast32_t additional_len(uint8_t additional)
 		 * 26 => 4
 		 * 27 => 8
 		 */
-		return 1 << (additional - ZCBOR_VALUE_IS_1_BYTE);
+		return 1U << (additional - ZCBOR_VALUE_IS_1_BYTE);
 	}
 	return 0;
 }
@@ -142,7 +142,7 @@ bool zcbor_int32_decode(zcbor_state_t *state, int32_t *result)
 		if (result64 > INT32_MAX) {
 			FAIL_RESTORE();
 		}
-		*result = result64;
+		*result = (int32_t)result64;
 		return true;
 	} else {
 		ZCBOR_FAIL();
@@ -714,7 +714,7 @@ bool zcbor_any_skip(zcbor_state_t *state, void *result)
 			temp_elem_count = state->elem_count;
 			state->elem_count = value;
 			if (!zcbor_multi_decode(value, value, &num_decode,
-					(void *)zcbor_any_skip, state,
+					(zcbor_decoder_t *)zcbor_any_skip, state,
 					NULL, 0)) {
 				state->elem_count = elem_count_bak;
 				state->payload = payload_bak;

--- a/src/zcbor_decode.c
+++ b/src/zcbor_decode.c
@@ -305,6 +305,11 @@ static bool str_start_decode(zcbor_state_t *state,
 
 bool zcbor_bstr_start_decode(zcbor_state_t *state, struct zcbor_string *result)
 {
+	if (result == NULL) {
+		struct zcbor_string dummy;
+		result = &dummy;
+	}
+
 	if(!str_start_decode(state, result, ZCBOR_MAJOR_TYPE_BSTR)) {
 		ZCBOR_FAIL();
 	}

--- a/src/zcbor_encode.c
+++ b/src/zcbor_encode.c
@@ -33,7 +33,7 @@ static uint8_t log2ceil(uint_fast32_t val)
 
 static uint8_t get_additional(uint_fast32_t len, uint8_t value0)
 {
-	return len == 0 ? value0 : (24 + log2ceil(len));
+	return len == 0 ? value0 : (uint8_t)(24 + log2ceil(len));
 }
 
 static bool encode_header_byte(zcbor_state_t *state,
@@ -45,7 +45,7 @@ static bool encode_header_byte(zcbor_state_t *state,
 
 	zcbor_assert(additional < 32, NULL);
 
-	*(state->payload_mut++) = (major_type << 5) | (additional & 0x1F);
+	*(state->payload_mut++) = (uint8_t)((major_type << 5) | (additional & 0x1F));
 	return true;
 }
 
@@ -103,7 +103,7 @@ static uint_fast32_t get_result_len(const void *const input, uint_fast32_t max_r
 	}
 
 	/* Round up to nearest power of 2. */
-	return len <= 2 ? len : (1 << log2ceil(len));
+	return len <= 2 ? len : (uint8_t)(1 << log2ceil(len));
 }
 
 
@@ -261,7 +261,8 @@ static bool primitive_put(zcbor_state_t *state, uint32_t input)
 static size_t remaining_str_len(zcbor_state_t *state)
 {
 	size_t max_len = (size_t)state->payload_end - (size_t)state->payload;
-	size_t result_len = get_result_len(&max_len, sizeof(size_t));
+	size_t result_len = get_result_len(&max_len, sizeof(max_len));
+
 	return max_len - result_len - 1;
 }
 
@@ -286,11 +287,11 @@ bool zcbor_bstr_start_encode(zcbor_state_t *state)
 bool zcbor_bstr_end_encode(zcbor_state_t *state)
 {
 	const uint8_t *payload = state->payload;
+	struct zcbor_string value;
 
 	if (!zcbor_process_backup(state, ZCBOR_FLAG_RESTORE | ZCBOR_FLAG_CONSUME, 0xFFFFFFFF)) {
 		ZCBOR_FAIL();
 	}
-	struct zcbor_string value;
 
 	value.value = state->payload_end - remaining_str_len(state);
 	value.len = (size_t)payload - (size_t)value.value;
@@ -405,7 +406,7 @@ static bool list_map_end_encode(zcbor_state_t *state, uint_fast32_t max_num,
 
 	if (max_header_len != header_len) {
 		const uint8_t *start = state->payload + max_header_len - header_len;
-		size_t body_size = payload - start;
+		size_t body_size = (size_t)payload - (size_t)start;
 		memmove(state->payload_mut,
 			state->payload + max_header_len - header_len,
 			body_size);
@@ -452,7 +453,7 @@ bool zcbor_undefined_put(zcbor_state_t *state, const void *unused)
 
 bool zcbor_bool_encode(zcbor_state_t *state, const bool *input)
 {
-	if (!primitive_put(state, *input + ZCBOR_BOOL_TO_PRIM)) {
+	if (!primitive_put(state, (uint32_t)(*input + ZCBOR_BOOL_TO_PRIM))) {
 		ZCBOR_FAIL();
 	}
 	return true;
@@ -461,7 +462,7 @@ bool zcbor_bool_encode(zcbor_state_t *state, const bool *input)
 
 bool zcbor_bool_put(zcbor_state_t *state, bool input)
 {
-	if (!primitive_put(state, input + ZCBOR_BOOL_TO_PRIM)) {
+	if (!primitive_put(state, (uint32_t)(input + ZCBOR_BOOL_TO_PRIM))) {
 		ZCBOR_FAIL();
 	}
 	return true;

--- a/tests/cmake/test_template.cmake
+++ b/tests/cmake/test_template.cmake
@@ -2,7 +2,7 @@
 cmake_minimum_required(VERSION 3.13.1)
 
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
-project(NONE)
+project(${PROJECT_NAME})
 
 FILE(GLOB app_sources src/*.c)
 target_sources(app PRIVATE

--- a/tests/decode/test1_suit_old_formats/CMakeLists.txt
+++ b/tests/decode/test1_suit_old_formats/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test1_suit_old_formats)
 include(../../cmake/test_template.cmake)
 
 set(py_command3

--- a/tests/decode/test2_suit/CMakeLists.txt
+++ b/tests/decode/test2_suit/CMakeLists.txt
@@ -31,3 +31,8 @@ include(${PROJECT_BINARY_DIR}/manifest2.cmake)
 
 target_link_libraries(manifest2 PRIVATE zephyr_interface)
 target_link_libraries(app PRIVATE manifest2)
+
+if (NOT VERBOSE)
+  # VERBOSE means including printk which doesn't build with these options.
+  target_compile_options(manifest2 PRIVATE -Wpedantic -Wconversion)
+endif()

--- a/tests/decode/test2_suit/CMakeLists.txt
+++ b/tests/decode/test2_suit/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test2_suit)
 include(../../cmake/test_template.cmake)
 
 set(py_command

--- a/tests/decode/test3_simple/CMakeLists.txt
+++ b/tests/decode/test3_simple/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test3_simple)
 include(../../cmake/test_template.cmake)
 
 target_sources(app PRIVATE

--- a/tests/decode/test5_strange/CMakeLists.txt
+++ b/tests/decode/test5_strange/CMakeLists.txt
@@ -47,14 +47,20 @@ execute_process(
   COMMAND_ERROR_IS_FATAL ANY
 )
 
-target_sources(app PRIVATE
+add_library(strange
   ${PROJECT_BINARY_DIR}/src/strange.c
   ${PROJECT_BINARY_DIR}/src/zcbor_decode.c
   ${PROJECT_BINARY_DIR}/src/zcbor_common.c)
+
+target_include_directories(strange PUBLIC ${PROJECT_BINARY_DIR}/include)
+target_link_libraries(strange PRIVATE zephyr_interface)
+target_link_libraries(app PRIVATE strange)
 
 if (TEST_INDETERMINATE_LENGTH_ARRAYS)
   target_compile_definitions(app PUBLIC TEST_INDETERMINATE_LENGTH_ARRAYS)
 endif()
 
-target_include_directories(app PRIVATE
-  ${PROJECT_BINARY_DIR}/include)
+if (NOT VERBOSE)
+  # VERBOSE means including printk which doesn't build with these options.
+  target_compile_options(strange PRIVATE -Wpedantic -Wconversion)
+endif()

--- a/tests/decode/test5_strange/CMakeLists.txt
+++ b/tests/decode/test5_strange/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test5_strange)
 include(../../cmake/test_template.cmake)
 
 set(py_command

--- a/tests/decode/test5_strange/src/main.c
+++ b/tests/decode/test5_strange/src/main.c
@@ -1295,7 +1295,6 @@ void test_main(void)
 			 ztest_unit_test(test_single),
 			 ztest_unit_test(test_unabstracted),
 			 ztest_unit_test(test_quantity_range),
-			 ztest_unit_test(test_unabstracted),
 			 ztest_unit_test(test_doublemap),
 			 ztest_unit_test(test_floats)
 	);

--- a/tests/decode/test7_suit9_simple/CMakeLists.txt
+++ b/tests/decode/test7_suit9_simple/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test7_suit9_simple)
 include(../../cmake/test_template.cmake)
 
 set(py_command

--- a/tests/decode/test8_suit12/CMakeLists.txt
+++ b/tests/decode/test8_suit12/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test8_suit12)
 include(../../cmake/test_template.cmake)
 
 set(py_command

--- a/tests/decode/test9_manifest14/CMakeLists.txt
+++ b/tests/decode/test9_manifest14/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test9_manifest14)
 include(../../cmake/test_template.cmake)
 
 if (NOT MANIFEST)

--- a/tests/encode/test1_suit/CMakeLists.txt
+++ b/tests/encode/test1_suit/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test1_suit)
 include(../../cmake/test_template.cmake)
 
 set(py_command_e

--- a/tests/encode/test2_simple/CMakeLists.txt
+++ b/tests/encode/test2_simple/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test2_simple)
 include(../../cmake/test_template.cmake)
 
 set(py_command

--- a/tests/encode/test2_simple/src/main.c
+++ b/tests/encode/test2_simple/src/main.c
@@ -153,7 +153,6 @@ void test_pet(void)
 		._Pet_name_tstr = {{.value = "foo", .len = 3}, {.value = "bar", .len = 3}},
 		._Pet_name_tstr_count = 2,
 		._Pet_birthday = {.value = (uint8_t[]){1,2,3,4,5,6,7,8}, .len = 8},
-		._Pet_species_dog = 2,
 		._Pet_species_choice = _Pet_species_dog
 	};
 	uint8_t exp_output[] = {
@@ -227,7 +226,8 @@ void test_pet_raw(void)
 	/* Check that encoding succeeded. */
 	zassert_true(res, NULL);
 	/* Check that the resulting length is correct. */
-	zassert_equal(sizeof(exp_output), state->payload - payload, NULL);
+	zassert_equal(sizeof(exp_output), state->payload - payload, "%d != %d\r\n",
+		sizeof(exp_output), state->payload - payload);
 	/* Check the payload contents. */
 	zassert_mem_equal(exp_output, payload, sizeof(exp_output), NULL);
 }

--- a/tests/encode/test3_strange/CMakeLists.txt
+++ b/tests/encode/test3_strange/CMakeLists.txt
@@ -48,3 +48,7 @@ include(${PROJECT_BINARY_DIR}/strange.cmake)
 target_link_libraries(strange PRIVATE zephyr_interface)
 target_link_libraries(app PRIVATE strange)
 
+if (NOT VERBOSE)
+  # VERBOSE means including printk which doesn't build with these options.
+  target_compile_options(strange PRIVATE -Wpedantic -Wconversion)
+endif()

--- a/tests/encode/test3_strange/CMakeLists.txt
+++ b/tests/encode/test3_strange/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test3_strange)
 include(../../cmake/test_template.cmake)
 
 set(py_command

--- a/tests/encode/test3_strange/src/main.c
+++ b/tests/encode/test3_strange/src/main.c
@@ -1145,7 +1145,6 @@ void test_main(void)
 			 ztest_unit_test(test_unabstracted),
 			 ztest_unit_test(test_string_overflow),
 			 ztest_unit_test(test_quantity_range),
-			 ztest_unit_test(test_unabstracted),
 			 ztest_unit_test(test_doublemap),
 			 ztest_unit_test(test_floats)
 	);

--- a/tests/test.sh
+++ b/tests/test.sh
@@ -25,4 +25,4 @@ if [[ -z "$ZEHPYR_BASE" ]]; then
         ZEPHYR_BASE=$(west topdir)/zephyr
 fi
 
-$ZEPHYR_BASE/scripts/twister -M -v -T . -W --platform native_posix --platform native_posix_64 $* -x VERBOSE=ON
+$ZEPHYR_BASE/scripts/twister -M -v -T . -W --platform native_posix --platform native_posix_64 $*

--- a/tests/unit/test1_unit_tests/CMakeLists.txt
+++ b/tests/unit/test1_unit_tests/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 cmake_minimum_required(VERSION 3.13.1)
 
+set(PROJECT_NAME test1_unit_tests)
 include(../../cmake/test_template.cmake)
 
 FILE(GLOB zcbor_sources ../../../src/*.c)

--- a/tests/verbose_test.sh
+++ b/tests/verbose_test.sh
@@ -1,0 +1,1 @@
+./test.sh -x VERBOSE=ON $*

--- a/zcbor/zcbor.py
+++ b/zcbor/zcbor.py
@@ -126,7 +126,7 @@ def val_or_null(value, var_name):
 def tmp_str_or_null(value):
     value_str = f'"{value}"' if value is not None else 'NULL'
     len_str = f"""sizeof({f'"{value}"'}) - 1, &tmp_str)"""
-    return f"(tmp_str.value = {value_str}, tmp_str.len = {len_str}"
+    return f"(tmp_str.value = (uint8_t *){value_str}, tmp_str.len = {len_str}"
 
 
 # Assign the max_value variable.
@@ -2270,15 +2270,16 @@ class CodeGenerator(CddlXcoder):
         if self.present_var_condition():
             func, *arguments = self.repeated_single_func(ptr_result=True)
             return (
-                f"zcbor_present_{self.mode}(&(%s), (void *)%s, %s)" %
+                f"zcbor_present_{self.mode}(&(%s), (zcbor_{self.mode}r_t *)%s, %s)" %
                 (self.present_var_access(),
                  func,
                  xcode_args(*arguments),))
         elif self.count_var_condition():
             func, *arguments = self.repeated_single_func(ptr_result=True)
-            minmax_str = "_minmax" if self.mode == "encode" else ""
+            minmax = "_minmax" if self.mode == "encode" else ""
+            mode = self.mode
             return (
-                f"zcbor_multi_{self.mode}{minmax_str}(%s, %s, &%s, (void *)%s, %s, %s)" %
+                f"zcbor_multi_{mode}{minmax}(%s, %s, &%s, (zcbor_{mode}r_t *)%s, %s, %s)" %
                 (self.min_qty,
                  self.max_qty,
                  self.count_var_access(),

--- a/zcbor/zcbor.py
+++ b/zcbor/zcbor.py
@@ -2150,9 +2150,9 @@ class CodeGenerator(CddlXcoder):
                             child.full_xcode(union_uint="DROP"))
                         for child in self.value])
                 bit_size = self.value[0].bit_size()
-                func = f"zcbor_uint{bit_size}_{self.mode}"
+                func = f"zcbor_uint32_{self.mode}"
                 return "((%s) && (%s))" % (
-                    f"({func}(state, (uint{bit_size}_t *)&{self.choice_var_access()}))",
+                    f"({func}(state, (uint32_t *)&{self.choice_var_access()}))",
                     f"{newl_ind}|| ".join(lines), )
             child_values = ["(%s && ((%s = %s) || 1))" %
                             (child.full_xcode(

--- a/zcbor/zcbor.py
+++ b/zcbor/zcbor.py
@@ -947,9 +947,11 @@ class CddlXcoder(CddlParser):
                 or (self.type == "OTHER" and self.my_types[self.value].is_unambiguous()))
 
     def is_unambiguous_repeated(self):
-        return self.is_unambiguous_value() \
-            and (self.key is None or self.key.is_unambiguous_repeated()) \
-            or (self.type in ["LIST", "GROUP", "MAP"] and len(self.value) == 0)
+        return (self.is_unambiguous_value()
+                and (self.key is None or self.key.is_unambiguous_repeated())
+                or (self.type in ["LIST", "GROUP", "MAP"] and len(self.value) == 0)
+                or (self.type in ["LIST", "GROUP", "MAP"]
+                    and all((child.is_unambiguous() for child in self.value))))
 
     def is_unambiguous(self):
         return (self.is_unambiguous_repeated() and (self.min_qty == self.max_qty))
@@ -1087,12 +1089,12 @@ class CddlXcoder(CddlParser):
             or self.cbor_var_condition()
             or self.tags
             or self.type_def_condition()
-            or (self.type in ["LIST", "MAP"] and len(self.value) != 0))
+            or (self.type in ["LIST", "MAP", "GROUP"] and len(self.value) != 0))
 
     # Whether this element needs its own encoder/decoder function.
     def repeated_single_func_impl_condition(self):
         return self.repeated_type_def_condition() \
-            or (self.type in ["LIST", "MAP"] and self.multi_member()) \
+            or (self.type in ["LIST", "MAP", "GROUP"] and self.multi_member()) \
             or (
                 self.multi_var_condition()
                 and (self.self_repeated_multi_var_condition() or self.range_check_condition()))
@@ -1671,9 +1673,11 @@ class CodeGenerator(CddlXcoder):
 
     # Declaration of the variables of all children.
     def child_single_declarations(self):
-        decl = [
-            line for child in self.value for line in child.add_var_name(
-                child.single_var_type(), anonymous=True)]
+        decl = list()
+        for child in self.value:
+            if not child.is_unambiguous_repeated():
+                decl.extend(child.add_var_name(
+                    child.single_var_type(), anonymous=True))
         return decl
 
     def simple_func_condition(self):
@@ -1793,7 +1797,10 @@ class CodeGenerator(CddlXcoder):
 
     # Enclose a list of declarations in a block (struct, union or enum).
     def enclose(self, ingress, declaration):
-        return [f"{ingress} {{"] + [indentation + line for line in declaration] + ["}"]
+        if declaration:
+            return [f"{ingress} {{"] + [indentation + line for line in declaration] + ["}"]
+        else:
+            return []
 
     # Type declaration for unions.
     def union_type(self):
@@ -1903,9 +1910,13 @@ class CodeGenerator(CddlXcoder):
         if self.type == "OTHER":
             ret_val.extend(self.my_types[self.value].type_def())
         if self.repeated_type_def_condition():
-            ret_val.extend([(self.single_var_type(full=False), self.repeated_type_name())])
+            type_def_list = self.single_var_type(full=False)
+            if type_def_list:
+                ret_val.extend([(self.single_var_type(full=False), self.repeated_type_name())])
         if self.type_def_condition():
-            ret_val.extend([(self.single_var_type(), self.type_name())])
+            type_def_list = self.single_var_type()
+            if type_def_list:
+                ret_val.extend([(self.single_var_type(), self.type_name())])
         return ret_val
 
     def type_def_bits(self):
@@ -2168,7 +2179,8 @@ class CodeGenerator(CddlXcoder):
 
     def xcode_bstr(self):
         if self.cbor_var_condition():
-            access_arg = f', &{self.val_access()}' if self.mode == 'decode' else ''
+            access_arg = f', {deref_if_not_null(self.val_access())}' if self.mode == 'decode' \
+                else ''
             xcode_cbor = "(%s)" % ((newl_ind + "&& ").join(
                 [f"(int_res = (zcbor_bstr_start_{self.mode}(state{access_arg})",
                  f"{self.cbor.full_xcode()})), zcbor_bstr_end_{self.mode}(state), int_res"]))

